File: docs/experiments/18-state-estimation/state_estimation.py
#!/usr/bin/env python3
"""Experiment 18 -- State estimation: small dense matrices in a long loop.

Runs the same algorithm as ``state_estimation.m``, in the same order, with the
same sizes.  See ``README.md`` for the measurements and the analysis.

    python3 state_estimation.py

WHAT IT MEASURES.  Every other benchmark in the suite is dominated by the cost
of ARRAYS.  This one is dominated by the cost of a CALL: 20000 iterations of
about ten operations on 6x6 and 2x2 matrices, for a total flop count a single
``dgemm`` would do in one call.

NumPy is not fast here either -- it is a Python loop around small BLAS calls,
which is exactly the regime the row is meant to expose.  It is nevertheless
the fastest of the three, and the ``README.md`` roadmap explains why and what
it would take for Mathilda to lead.

DETERMINISM.  The measurement sequence is a fixed pair of sinusoids, so
``trace(P)`` at the end is an exact cross-system check.  The filter is stable,
so P converges to a steady state rather than drifting.
"""

import math
import time

import numpy as np

# ---- shared reporting helpers (identical in every experiment file) --------


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))


# ---- model ---------------------------------------------------------------

KFN = 20000                # measurements
KFDT = 0.01                # sample interval

# Constant-acceleration transition in 2D: state is {x, y, vx, vy, ax, ay}.
kfF = np.zeros((6, 6))
for i in range(6):
    kfF[i, i] = 1.0
    if i < 4:
        kfF[i, i + 2] = KFDT
    if i < 2:
        kfF[i, i + 4] = 0.5 * KFDT ** 2
kfFT = kfF.T.copy()

kfH = np.zeros((2, 6))                 # observe x, y
kfH[0, 0] = 1.0
kfH[1, 1] = 1.0
kfHT = kfH.T.copy()
kfQ = 0.001 * np.eye(6)                # process noise
kfR = 0.05 * np.eye(2)                 # measurement noise
kfI = np.eye(6)

kfx0 = np.zeros(6)
kfP0 = np.eye(6)


def kfz(k):
    """A deterministic 'trajectory plus sensor noise'."""
    return np.array([math.sin(0.01 * k) + 0.02 * math.sin(7.1 * k),
                     math.cos(0.013 * k) + 0.02 * math.cos(5.3 * k)])


# ---- 1. the scalar filter ------------------------------------------------


def kfstep(st, k):
    """Textbook predict/update.  Every matrix is 6x6 or smaller."""
    xp = kfF @ st[0]                       # predict state
    pp = kfF @ st[1] @ kfFT + kfQ          # predict covariance
    yy = kfz(k) - kfH @ xp                 # innovation
    ss = kfH @ pp @ kfHT + kfR             # innovation covariance, 2x2
    kk = pp @ kfHT @ np.linalg.inv(ss)     # Kalman gain, 6x2
    return [xp + kk @ yy, (kfI - kk @ kfH) @ pp]


def kalman(m):
    st = [kfx0, kfP0]
    k = 1
    while k <= m:
        st = kfstep(st, k)
        k += 1
    return float(np.trace(st[1]))


# ---- 2. the ensemble filter ----------------------------------------------

ENN = 4096                 # ensemble members
ENS = 200                  # steps

enE0 = 0.01 * np.sin(np.arange(1, ENN + 1)[:, None] + 6 * np.arange(1, 7)[None, :])


def enstep(e, k):
    """The same estimator, array-shaped: the covariance is estimated from the
    spread of the ensemble rather than propagated from a model.  One 4096x6
    GEMM replaces the 6x6 algebra, so this row lives in the opposite regime."""
    pr = e @ kfFT                          # advance every member
    mn = pr.sum(0) / ENN                   # ensemble mean
    an = pr - mn[None, :]                  # anomalies
    pc = an.T @ an / (ENN - 1.0)           # sample covariance, 6x6
    kk = pc @ kfHT @ np.linalg.inv(kfH @ pc @ kfHT + kfR)
    inn = np.tile(kfz(k), (ENN, 1)) - pr @ kfHT
    return pr + inn @ kk.T


def enkf(m):
    e = enE0
    k = 1
    while k <= m:
        e = enstep(e, k)
        k += 1
    return float((e.sum(0) / ENN).sum())


def main():
    print("Experiment 18 -- state estimation")
    print("")

    bench("Kalman filter, 6 states, 20000 steps", lambda: kalman(KFN))
    check("Kalman filter (2000 steps)", kalman(2000))

    bench("ensemble Kalman, 4096 members, 200 steps", lambda: enkf(ENS))
    check("ensemble Kalman (20 steps)", enkf(20))

    # What one small dense operation costs.  NumPy pays Python-call overhead
    # here rather than kernel time, which is the honest comparison for the
    # .m file's numbers: both systems are measuring dispatch, not flops.
    print("")
    print("-- what one small dense operation costs --")
    sm6 = np.full((6, 6), 0.3) + 1.7 * np.eye(6)
    sv6 = np.arange(1.0, 7.0)
    sm2 = np.array([[2.0, 0.3], [0.3, 1.7]])
    sm60 = np.full((60, 60), 0.3) + 59.7 * np.eye(60)

    def rep(f, n):
        return lambda: [f() for _ in range(n)]

    bench("inv(2x2) x 20000", rep(lambda: np.linalg.inv(sm2), 20000))
    bench("inv(6x6) x 2000", rep(lambda: np.linalg.inv(sm6), 2000))
    bench("m6 @ v6  x 50000", rep(lambda: sm6 @ sv6, 50000))
    bench("m6 @ m6  x 20000", rep(lambda: sm6 @ sm6, 20000))
    bench("inv(60x60) x 2000", rep(lambda: np.linalg.inv(sm60), 2000))


if __name__ == "__main__":
    main()

File: docs/experiments/18-state-estimation/state_estimation.m
(* ==========================================================================
   Experiment 18 -- State estimation: small dense matrices in a long loop
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file state_estimation.m
       Mathematica   wolframscript -file state_estimation.m
       Python        python3 state_estimation.py

   WHAT IT MEASURES.  Every other benchmark in the suite is dominated by the
   cost of ARRAYS.  This one is dominated by the cost of a CALL.

   A 6-state constant-acceleration Kalman filter over 20000 measurements does,
   per step, two 6x6 products, a 6x6-by-6 matrix-vector, a 2x6-by-6x6, a 2x2
   inverse and half a dozen small adds.  The flop count for the whole run is
   about 2e7 -- a single dgemm in the linear-algebra suite does more than that
   in one call.  What is being measured is 20000 iterations x ~10 operations
   of per-operation dispatch on data too small for any kernel to matter.

   NumPy is not fast here either, and that is the point: the row shows where
   all three systems live when the data is small.  The ensemble filter beside
   it is the same estimator written the array way -- 4096 members at once --
   so the two regimes can be read together.

   DETERMINISM.  The measurement sequence is a fixed pair of sinusoids, so
   Tr[P] at the end is an exact cross-system check.  The filter is stable, so
   P converges to a steady state rather than drifting, which makes the check
   insensitive to the last-bit differences three systems will always have.
   ========================================================================== *)

(* ---- shared reporting helpers (identical in every experiment file) ------- *)

SetAttributes[bench, HoldRest];

bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];

(* ---- model -------------------------------------------------------------- *)

kfn  = 20000;      (* measurements                                          *)
kfdt = 0.01;       (* sample interval                                       *)

(* Constant-acceleration transition in 2D: state is
   {x, y, vx, vy, ax, ay}, so position picks up v dt + a dt^2/2 and velocity
   picks up a dt. *)
kfF = Table[Which[i == j,                 1.,
                  j == i + 2 && i <= 4,   kfdt,
                  j == i + 4 && i <= 2,   0.5 kfdt^2,
                  True,                   0.], {i, 1, 6}, {j, 1, 6}];
kfFT = Transpose[kfF];

kfH  = Table[If[i == j, 1., 0.], {i, 1, 2}, {j, 1, 6}];   (* observe x, y   *)
kfHT = Transpose[kfH];
kfQ  = Table[If[i == j, 0.001, 0.], {i, 1, 6}, {j, 1, 6}]; (* process noise *)
kfR  = Table[If[i == j, 0.05,  0.], {i, 1, 2}, {j, 1, 2}]; (* meas. noise   *)
kfI  = Table[If[i == j, 1., 0.], {i, 1, 6}, {j, 1, 6}];

kfx0 = Table[0., {6}];
kfP0 = Table[If[i == j, 1., 0.], {i, 1, 6}, {j, 1, 6}];

(* A deterministic "trajectory plus sensor noise". *)
kfz[k_] := {Sin[0.01 k] + 0.02 Sin[7.1 k], Cos[0.013 k] + 0.02 Cos[5.3 k]};

(* ---- 1. the scalar filter ----------------------------------------------- *)

(* Textbook predict/update.  Every matrix here is 6x6 or smaller, so none of
   them is anywhere near the packing threshold -- which is the finding this
   experiment exists to make; see README.md. *)
kfstep[st_, k_] := Module[{xp, pp, yy, ss, kk},
  xp = kfF . st[[1]];                       (* predict state                *)
  pp = kfF . st[[2]] . kfFT + kfQ;          (* predict covariance           *)
  yy = kfz[k] - kfH . xp;                   (* innovation                   *)
  ss = kfH . pp . kfHT + kfR;               (* innovation covariance, 2x2   *)
  kk = pp . kfHT . Inverse[ss];             (* Kalman gain, 6x2             *)
  {xp + kk . yy, (kfI - kk . kfH) . pp}];

kalman[m_] := Module[{st, k},
  st = {kfx0, kfP0}; k = 1;
  While[k <= m, st = kfstep[st, k]; k = k + 1];
  Tr[st[[2]]]];

(* ---- 2. the ensemble filter --------------------------------------------- *)

enn = 4096;        (* ensemble members                                      *)
ens = 200;         (* steps                                                 *)

enE0 = Table[0.01 Sin[1. (i + 6 j)], {i, 1, enn}, {j, 1, 6}];

(* The same estimator, array-shaped: the covariance is never formed from a
   model, it is estimated from the spread of the ensemble.  One 4096x6 GEMM
   replaces the 6x6 algebra, so this row lives in the opposite regime to the
   one above. *)
enstep[e_, k_] := Module[{pr, mn, an, pc, kk, inn},
  pr  = e . kfFT;                           (* advance every member         *)
  mn  = Total[pr]/enn;                      (* ensemble mean                *)
  an  = pr - Table[mn, {enn}];              (* anomalies                    *)
  pc  = Transpose[an] . an/(enn - 1.);      (* sample covariance, 6x6       *)
  kk  = pc . kfHT . Inverse[kfH . pc . kfHT + kfR];
  inn = Table[kfz[k], {enn}] - pr . kfHT;
  pr + inn . Transpose[kk]];

enkf[m_] := Module[{e, k},
  e = enE0; k = 1;
  While[k <= m, e = enstep[e, k]; k = k + 1];
  Total[Total[e]/enn]];

(* ---- run ---------------------------------------------------------------- *)

Print["Experiment 18 -- state estimation"];
Print[""];

bench["Kalman filter, 6 states, 20000 steps",     kalman[kfn]];
check["Kalman filter (2000 steps)", kalman[2000]];

bench["ensemble Kalman, 4096 members, 200 steps", enkf[ens]];
check["ensemble Kalman (20 steps)", enkf[20]];

(* ---- what a small matrix costs ------------------------------------------ *)
(*
   A 6x6 matrix is 36 elements, far below the packing threshold, so it is
   correctly never packed -- and Inverse then cannot reach the numeric path at
   all and runs a fraction-free symbolic elimination instead.  These are the
   numbers behind that claim.  For scale: a 6x6 dgetrf+dgetri is about 1
   microsecond.
*)
Print[""];
Print["-- what one small dense operation costs --"];

sm6 = Table[If[i == j, 2., 0.3], {i, 1, 6}, {j, 1, 6}];
sv6 = Table[N[i], {i, 1, 6}];
sm2 = {{2., 0.3}, {0.3, 1.7}};

bench["Inverse[2x2] x 20000",   Do[Inverse[sm2], {20000}]];
bench["Inverse[6x6] x 2000",    Do[Inverse[sm6], {2000}]];
bench["m6 . v6      x 50000",   Do[sm6 . sv6, {50000}]];
bench["m6 . m6      x 20000",   Do[sm6 . sm6, {20000}]];

(* For contrast, the same operation on data ABOVE the packing threshold,
   where it reaches LAPACK.  Per element this is orders of magnitude cheaper. *)
sm60 = Table[If[i == j, 60., 0.3], {i, 1, 60}, {j, 1, 60}];
bench["Inverse[60x60] x 2000 (above threshold)", Do[Inverse[sm60], {2000}]];
